homeassistant/components/opower/coordinator.py

```diff
@@ -156,32 +169,100 @@ async def _insert_statistics(self) -> None:
                 cost_sum = cast(float, stats[cost_statistic_id][0]["sum"])
                 consumption_sum = cast(float, stats[consumption_statistic_id][0]["sum"])
                 last_stats_time = stats[consumption_statistic_id][0]["start"]
+                if cost_reads_billing:
+                    start = cost_reads_billing[0].start_time
+                    _LOGGER.debug("Getting cost statistics at: %s", start)
+                    # In the common case there should be a previous statistic at start time
+                    # so we only need to fetch one statistic. If there isn't any, fetch all.
+                    for end in (start + timedelta(seconds=1), None):
+                        stats = await get_instance(self.hass).async_add_executor_job(
+                            statistics_during_period,
+                            self.hass,
+                            start,
+                            end,
+                            {cost_statistic_id},
+                            "hour",
+                            None,
+                            {"sum"},
+                        )
+                        if stats:
+                            break
+                        if end:
+                            _LOGGER.debug(
+                                "Not found. Trying to find the oldest statistic after %s",
+                                start,
+                            )
+                    # We are in this code path only if get_last_statistics found a stat
+                    # so statistics_during_period should also have found at least one.
+                    assert stats
+                    cost_sum = cast(float, stats[cost_statistic_id][0]["sum"])
+                    last_stats_time_billing = stats[cost_statistic_id][0]["start"]
 
-            cost_statistics = []
-            consumption_statistics = []
+            await self._insert_statistics_cost_read(
+                account,
+                cost_statistic_id,
+                consumption_statistic_id,
+                cost_sum,
+                consumption_sum,
+                last_stats_time,
+                cost_reads,
+                [StatisticType.COST, StatisticType.CONSUMPTION]
```

Comment:
The code is now hard to maintain with all the branching. I feel like it could be simplified if you have 2 lists of cost reads: one for usage, one for cost. For accounts like mine the 2 lists will be the same. For accounts like yours the second list will have the monthly data. Then you can fetch and update statistics in the recorder separately. This unfortunately means for accounts like mine the recorder interactions will now be slower but I think it's better to have less complex code. Could you try that in a separate PR to see how the code would look like?